src/pgn.js

```diff
@@ -337,9 +355,334 @@ exports.pgnRead = function(pgnString, gameIndex) {
 				++gameCounter;
 			}
 			else {
-				throw new exception.InvalidPGN(pgnString, pgnString.length, i18n.INVALID_GAME_INDEX, gameIndex, gameCounter);
+				throw new exception.InvalidPGN(pgnString, pgnString.length, stream._lineCount, i18n.INVALID_GAME_INDEX, gameIndex, gameCounter);
 			}
 		}
 		return doParseGame(stream);
 	}
 };
+
+/**
+ * Write the date in PGN format
+ *
+ * @param {Date|{year:number, month:number}|{year:number}|undefined}} date
+ *
+ * @returns {string}
+ */
+function writeDate(date) {
+	if (date === undefined) {
+		return '[Date \'????.??.??\']' + Separator;
+	} else if (date instanceof Date) {
+		return '[Date \'' +
+			date.getFullYear().toString() + '.' +
+			((date.getMonth() + 1 < 10) ? ('0' + (date.getMonth() + 1).toString()) : (date.getMonth() + 1)).toString() + '.' +
+			((date.getDate() < 10) ? ('0' + date.getDate().toString()) : date.getDate().toString()) +
+			'\']' + Separator;
+	} else {
+		return '[Date \'' +
+			((date.year === undefined) ? '????' : date.year) + '.' +
+			((date.month === undefined) ? '??' : date.month) + '.' +
+			((date.date === undefined) ? '??' : date.date) +
+			'\']' + Separator;
+	}
+}
+
+/**
+ * Write non Seven Tag Roster tags
+ * @param {Game} game
+ *
+ * @returns {string}
+ */
+function writeNonSTRTags(game) {
+	var res = '';
+	var STR = ['Event', 'Site', 'Date', 'Round', 'White', 'Black', 'Result'];
+	if (Object.keys(game.tags()).length > 0) {
+		var tags = game.tags();
+		for (var tag in game.tags()) {
+			if (Object.prototype.hasOwnProperty.call(tags, tag)) {
+				var value = tags[tag];
+				if (!STR.includes(tag)) {
+					res += ('[' + tag + ' \'' + value + '\']' + Separator);
+				}
+			}
+		}
+		/*Object.entries(game.tags()).forEach((tag, value) => {
+			if (!STR.includes(tag)) {
+				res += ('[' + tag + ' \'' + value + '\']' + Separator);
+			}
+		});*/
+	}
+	return res;
+}
+
+/**
+ * Write the headers in order, first the standard ones, then any others
+ *
+ * @param {Game} game
+ *
+ * @returns {string}
+ */
+function writeHeaders(game) {
+	var res = '';
+	res += '[Event \'' + ((game.event() === undefined) ? '?' : game.event()) + '\']' + Separator;
```

Comment:
Tag values must be espaced. See unescaping method `parseHeaderValue(..)` in tokenstream.js.

Reply:
Will do, expect I'll have a couple more revisions coming in the next few days as I also want to create a "MainVariation" key in the JSON even though it would duplicate data as it would make it much cleaner for doing position search, material search etc.